### PR TITLE
Speeding up Array_Proxy.from_proxy_object twice

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Using lambda instead of higher order function.

### Important Notes

Running:
```
sbt:runtime> benchOnly VectorBenchmarks.averageOverArrayProxy
```
speeds up from `0.038 ms/op` to `0.016 ms/op` on my computer. Which seems good enough.

### Checklist

- All code has been tested:
  - [x] Benchmark is improved

### CODE REVIEW

**Review by radeusgd — APPROVED** [review]
Cool!

But what happens here? Shouldn't both have the same speed?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java [hunk 1]
```diff
@@ -59,13 +63,16 @@ boolean isArrayElementReadable(long index) {
...
+      @Cached("create()") BranchProfile arrayIndexHasHappened,
+      @CachedLibrary(limit = "3") InteropLibrary interop)
       throws UnsupportedMessageException, InvalidArrayIndexException {
-    if (index >= length || index < 0) {
-      throw InvalidArrayIndexException.create(index);
-    }
-
     try {
+      if (index >= length || index < 0) {
+        arrayIndexHasHappened.enter();
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java:73]: This profile gets me from 20us to 16us on my computer.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array_Proxy.enso [hunk 2]
```diff
@@ -33,4 +33,4 @@ type Array_Proxy
        methods.
     from_proxy_object : Any -> Array
     from_proxy_object proxy =
-        Array_Proxy.new proxy.length proxy.at
+        Array_Proxy.new proxy.length (i -> proxy.at i)
```
Comment (hubertp) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array_Proxy.enso:36]: This is it? ~Don'w we do eta-expansion at the compiler?~ And to answer my own question, no. I think I start to understand where the problem is. I also think there are plenty of similar situations in stdlib

**Review by hubertp — APPROVED** [review]

**Comment by JaroslavTulach** [comment]
> But what happens here? Shouldn't both have the same speed?

Yes, they should. I believe the difference is somewhere here:

image

The lambda directly invokes `callTarget`, but the _"curried function"_ does some `Object[].clone()` - that's a visible (e.g. not virtual) memory allocation and I believe it slows the invocation down. CCing @kustosz

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array_Proxy.enso (hunk 2 above)
Comment (JaroslavTulach) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array_Proxy.enso:36]: Based on [comparing the graphs](https://user-images.githubusercontent.com/26887752/206692254-37169384-b474-4b80-8dc3-ace5891c4943.png) I can speed the original case also by:
```diff
--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/function/Function.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/function/Function.java
@@ -159,7 +159,7 @@
    * @return a copy of the partially applied arguments for this function
    */
   public Object[] clonePreAppliedArguments() {
-    return preAppliedArguments.clone();
+    return preAppliedArguments;
   }
 
   /**
```
Reported as [pivotal-184001258](https://www.pivotaltracker.com/n/projects/2539304/stories/184001258).

**Review by 4e6 — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array_Proxy.enso (hunk 2 above)
Comment (kustosz) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array_Proxy.enso:36]: If I were to guess, the original would be faster.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java (hunk 1 above)
Comment (kustosz) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java:73]: That makes sense.  Why is it pulled into the try though?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java [hunk 3]
```diff
@@ -59,13 +63,16 @@ boolean isArrayElementReadable(long index) {
   }
 
   @ExportMessage
-  public Object readArrayElement(long index, @CachedLibrary(limit = "3") InteropLibrary interop)
+  public Object readArrayElement(
+      long index,
+      @Cached("create()") BranchProfile arrayIndexHasHappened,
```
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java:68]: `@Cached` will do, you don't need the `create()`, that's the default.

**Review by kustosz — APPROVED** [review]

**Comment by kustosz** [comment]
Interesting find about the `clone`. I don't think we can just remove it, that would make programs incorrect. But I cannot help but wonder if there's a way to virtualize that? Maybe use `System.arraycopy` or an exploded loop instead? I see no reason why the arguments would be allocated other than "Truffle got confused". The fact that there exists an `ObjectClone` node would suggest I'm right – i.e. using `Object[].clone` compiles differently than e.g. `System.arraycopy`. Can you try that please?

**Pull request merged**